File: 199.二叉树的右视图.js
/*
 * @lc app=leetcode.cn id=199 lang=javascript
 *
 * [199] 二叉树的右视图
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var rightSideView = function (root) {
  let queue = [],
    front,
    rear,
    res = [];
  front = rear = 0;
  queue[rear++] = root;
  while (front !== rear) {
    let size = rear - front;
    let node;
    for (let i = 0; i < size; i++) {
      node = queue[front++];
      if (node && node.left) queue[rear++] = node.left;
      if (node && node.right) queue[rear++] = node.right;
    }
    if (node) res.push(node.val);
  }
  return res;
};
// @lc code=end

File: store/199.二叉树的右视图.c
/*
 * @lc app=leetcode.cn id=199 lang=c
 *
 * [199] 二叉树的右视图
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int *rightSideView(struct TreeNode *root, int *returnSize)
{
  int front, rear, *result, top, size, i;
  struct TreeNode **queue, *temp;
  queue = (struct TreeNode **)malloc(sizeof(struct TreeNode *) * 100);
  result = (int *)malloc(sizeof(int) * 100);
  front = rear = top = 0;
  if (root != NULL)
  {
    queue[rear++] = root;
    while (front != rear)
    {
      size = rear;
      for (i = front; i < size; i++)
      {
        temp = queue[front++];
        if (temp->left != NULL)
          queue[rear++] = temp->left;
        if (temp->right != NULL)
          queue[rear++] = temp->right;
      }
      result[top++] = temp->val;
    }
  }
  *returnSize = top;
  return result;
}
// @lc code=end
